tsdb/engine/tsm1/file_store_test.go

```diff
@@ -3043,6 +3045,54 @@ func TestFileStore_Observer(t *testing.T) {
 	unlinks, finishes = nil, nil
 }
 
+func TestFileStore_PurgerContention(t *testing.T) {
+	t.Helper()
+
+	dir := MustTempDir()
+	defer os.RemoveAll(dir)
+
+	// Create some TSM files...
+	data := make([]keyValues, 0, 10)
+	for i := 0; i < 1000; i++ {
+		data = append(data, keyValues{"cpu", []tsm1.Value{tsm1.NewValue(0, 1.0)}})
+	}
+
+	regFiles, err := newFileDir(dir, data...)
+	require.NoError(t, err)
+	tmpFiles, err := newTmpFileDir(dir, data...)
+	require.NoError(t, err)
+
+	fs := tsm1.NewFileStore(dir)
+	err = fs.Open()
+	require.NoError(t, err)
+
+	done := make(chan struct{})
+
+	go func() {
+		var overlapsTimeRangeMinMaxLock sync.Mutex
+		fs.Apply(func(r tsm1.TSMFile) error {
+			overlapsTimeRangeMinMaxLock.Lock()
+			time.Sleep(time.Duration(rand.Intn(5)) * time.Millisecond)
+			overlapsTimeRangeMinMaxLock.Unlock()
+			return nil
+		})
```

Comment:
The use of `fs.Apply` to inject mischief is really smart. However, the `overlapsTimeRangeMinMaxLock` just makes the goroutines launched by `fs.Apply` run in a serial manner, equivalent to having the limiter in `fs.Apply` only allow 1 goroutine at a time. Given that, I'm not the mutex helps create contention more than just doing `time.Sleep`.